### PR TITLE
Core/Quest: Fix repeatable quest mark

### LINKED ISSUE #23087: Core/Quest: repeatable quest mark
**Description:** For a repeatable quest the mark must always be a "Blue ?"
The issue here is the quest mark is shown as (yellow !) if the quest level it's not "gray" (player level - quest level=<10) and it will turns  "Blue ?" once "gray" (player level - quest level>10) 

**Expected behavior:** The quest mark must always be  a "Blue ?"

**Steps to reproduce the problem:**

1. Connect a horde char level 80
2. .go c id 14727
3. complete all quests proposed until the repeatable quest with a "Blue ?" is displayed
4. .level -10 your char (level 70)
5. .die and .resp the npc
6. Notice the quest mark is displayed as (yellow !)
7. .level 1 (level 71)
8. .die and .resp the npc
9. Notice the quest mark is displayed as (Blue ?)

The quest=7837  Has Level: 60 it means it will turn gray at level 71

**Branch(es):** 3.3.5

**TC rev. hash/commit:** rev. 72052aa9a556

### PR DESCRIPTION
**Pull request by jackpoz**
**Changes proposed:**

-  Don't show a yellow ! for repeatable quests, even if they are still in level range not to be low level.

**Target branch(es):** 3.3.5/master

- [x] 3.3.5
- [ ] master

**Issues addressed:** Closes #  (insert issue tracker number)
- Closes #23087

**Tests performed:** (Does it build, tested in-game, etc.)
- Steps described in #23087

**Known issues and TODO list:** (add/remove lines as needed)

- [x] Should other conditions be added, like quest->IsDailyOrWeekly() or quest->IsAutoComplete() ?
- [x] Need to test other cases to see there is no regression

### CODE REVIEW

**Comment by Killyana** [comment]
There's also an issue with daily quests: https://www.wowhead.com/npc=25580/old-man-barlo#starts
1. If your char didn't learned fishing .cast 7733
2. .go c id 25580
3. Notice the yellow quest mark

Before this PR the mark was blue for char 70-80 and yellow for 81+
It must be blue whatever the level

**Comment by jackpoz** [comment]
PR updated, try now

**Comment by Killyana** [comment]
The quest mark must be a blue "!" and not a blue "?" for daily quests, otherwise it looks fine, I think I tested all possible cases

**Comment by Killyana** [comment]
I noticed also a difference between TC and retail in auto complete quests: https://classic.wowhead.com/quest=6383/the-ashenvale-hunt
image

**Comment by jackpoz** [comment]
Isn't a blue ! for dailies ? It showed like that for the fishing quest to me

**Comment by MaxtorCoder** [comment]
Yes, it is supposed to be blue 😄

**Comment by Killyana** [comment]
After further investigation:
-- repeatable with `questtype`=0 must have a blue "?"
-- repeatable with `questtype`=2  must have a yellow "!"
-- daily must have blue "!"
-- weekly must have yellow "!"

Issues:
-- Daily quest mark turns yellow if there's a +11 level difference between quest requirement and player level
-- Repeatable with `questtype`=0  turns yellow "!"  if the level difference between quest requirement and player level is less than +21 (when the quest turns gray the correct quest mark is displayed)

**Comment by illfated** [comment]
Are any low-level quests affected by this, like the repeatable quests to hand in cloth to the faction cloth quartermasters, or Chen's empty keg?

**Comment by Killyana** [comment]
I added some examples.

**Comment by jackpoz** [comment]
this issue got messy quite fast xD

**Comment by jackpoz** [comment]
PR updated with the changes above, still need to fix the gossip icon

**Comment by Wyrserth** [comment]
@jackpoz this fixes the gossip icon:

```diff
diff --git a/src/server/game/Entities/Creature/GossipDef.cpp b/src/server/game/Entities/Creature/GossipDef.cpp
index 6a8da9a82f..b29684ad42 100644
--- a/src/server/game/Entities/Creature/GossipDef.cpp
+++ b/src/server/game/Entities/Creature/GossipDef.cpp
@@ -224,10 +224,13 @@ void PlayerMenu::SendGossipMenu(uint32 titleTextId, ObjectGuid objectGUID)
         {
             ++count;
             data << uint32(questID);
-            data << uint32(item.QuestIcon);
+            data << (quest->IsAutoComplete() ? 0 : item.QuestIcon);
             data << int32(quest->GetQuestLevel());
             data << uint32(quest->GetFlags());              // 3.3.3 quest flags
-            data << uint8(0);                               // 3.3.3 changes icon: blue question or yellow exclamation
+            bool changeIcon = 0;
+            if (quest->IsAutoComplete() && quest->IsRepeatable())
+                changeIcon = true;
+            data << uint8(changeIcon);                      // 3.3.3 changes icon: blue question or yellow exclamation
             std::string title = quest->GetTitle();
 
LocaleConstant localeConstant = _session->GetSessionDbLocaleIndex();
```

**Comment by Killyana** [comment]
The quest The Ashenvale Hunt is auto-complete but not repeatable, if I remove " && quest->IsRepeatable()  " from the check the quest mark turns into a blue  "?" instead a yellow "!"

**Comment by Wyrserth** [comment]
@Killyana I recall that quest is supposed to have a yellow ! in both the quest icon inside the gossip and on top of the questgiver's head, so it should be fine as is?

**Comment by Killyana** [comment]
Yes, but in the actual revision it has a gray "?" inside the gossip
this fix all issues 
``` diff 
diff --git a/src/server/game/Entities/Creature/GossipDef.cpp b/src/server/game/Entities/Creature/GossipDef.cpp
index 6a8da9a82f..4dda1de7ec 100644
--- a/src/server/game/Entities/Creature/GossipDef.cpp
+++ b/src/server/game/Entities/Creature/GossipDef.cpp
@@ -224,10 +224,13 @@ void PlayerMenu::SendGossipMenu(uint32 titleTextId, ObjectGuid objectGUID)
         {
             ++count;
             data << uint32(questID);
-            data << uint32(item.QuestIcon);
+            data << (quest->IsAutoComplete() ? 0 : item.QuestIcon);
             data << int32(quest->GetQuestLevel());
             data << uint32(quest->GetFlags());              // 3.3.3 quest flags
-            data << uint8(0);                               // 3.3.3 changes icon: blue question or yellow exclamation
+            bool changeIcon = false;
+            if (quest->IsRepeatable())
+                changeIcon = true;
+            data << uint8(changeIcon);                      // 3.3.3 changes icon: blue question or yellow exclamation
             std::string title = quest->GetTitle();
 
             LocaleConstant localeConstant = _session->GetSessionDbLocaleIndex();

**Comment by Wyrserth** [comment]
Tested both on a few quests, indeed the second quest->IsAutoComplete() was useless. Nice catch!

**Comment by jackpoz** [comment]
I applied the patch above to the PR

**Comment by Killyana** [comment]
The patch in Wyrserth must be applied, I tested it and every thing looks fine.

**Comment by jackpoz** [comment]
I applied yours at https://github.com/TrinityCore/TrinityCore/pull/23338#issuecomment-498062002 , the one without AutoComplete(), isn't that the correct one ?

**Comment by Killyana** [comment]
No it has an issue with som quests this one works fine: https://github.com/TrinityCore/TrinityCore/pull/23338#issuecomment-498059105

**Comment by jackpoz** [comment]
PR updated

**Comment by Killyana** [comment]
We forget the low level check, basically the issue is the one who made the thing to hide low level quests used the same mark for all kind of quests,     DIALOG_STATUS_LOW_LEVEL_REWARD_REP     = 3,
and    DIALOG_STATUS_LOW_LEVEL_AVAILABLE_REP  = 4 aren't used.
This is the last fix:
``` diff
                            result2 = DIALOG_STATUS_AVAILABLE_REP;
                            if (getLevel() <= (GetQuestLevel(quest) + sWorld->getIntConfig(CONFIG_QUEST_LOW_LEVEL_HIDE_DIFF)))
                                result2 = DIALOG_STATUS_AVAILABLE_REP;
                            else
                                result2 =  DIALOG_STATUS_LOW_LEVEL_AVAILABLE_REP;
                        else if (quest->IsWeekly())
                            if (getLevel() <= (GetQuestLevel(quest) + sWorld->getIntConfig(CONFIG_QUEST_LOW_LEVEL_HIDE_DIFF)))
                                result2 = DIALOG_STATUS_AVAILABLE;
                            else
                                result2 =  DIALOG_STATUS_LOW_LEVEL_AVAILABLE;
                        else if (quest->IsAutoComplete())
                            if (getLevel() <= (GetQuestLevel(quest) + sWorld->getIntConfig(CONFIG_QUEST_LOW_LEVEL_HIDE_DIFF)))
                                result2 = DIALOG_STATUS_REWARD_REP;
                            else
                                DIALOG_STATUS_LOW_LEVEL_REWARD_REP;
                        else
                            result2 = DIALOG_STATUS_AVAILABLE;
                            if (getLevel() <= (GetQuestLevel(quest) + sWorld->getIntConfig(CONFIG_QUEST_LOW_LEVEL_HIDE_DIFF)))
                                result2 = DIALOG_STATUS_AVAILABLE;
                            else
                                result2 =  DIALOG_STATUS_LOW_LEVEL_AVAILABLE;
                    }
                    else if (getLevel() <= (GetQuestLevel(quest) + sWorld->getIntConfig(CONFIG_QUEST_LOW_LEVEL_HIDE_DIFF)))
                    {
                         result2 = DIALOG_STATUS_AVAILABLE;
                    }
                    else
                        result2 = DIALOG_STATUS_LOW_LEVEL_AVAILABLE;

File: src/server/game/Entities/Creature/GossipDef.cpp [hunk 1]
```diff
@@ -224,10 +224,13 @@ void PlayerMenu::SendGossipMenu(uint32 titleTextId, ObjectGuid objectGUID)
         {
             ++count;
             data << uint32(questID);
-            data << uint32(item.QuestIcon);
+            data << (quest->IsAutoComplete() ? 0 : item.QuestIcon);
```
Comment (Shauren) [inline src/server/game/Entities/Creature/GossipDef.cpp]: don't break packet structure - explicitly state the type

Comment (Shauren) [reply src/server/game/Entities/Creature/GossipDef.cpp]: also, cross-referencing master branch https://github.com/TrinityCore/TrinityCore/blob/master/src/server/game/Entities/Creature/GossipDef.cpp#L243

Gives you an idea what client expects in this packet - id, icon, level, flags, repeatable

Comment (jackpoz) [reply src/server/game/Entities/Creature/GossipDef.cpp]: I see there is text.Repeatable = quest->IsRepeatable(), while this PR has a check about IsAutoComplete() too

**Comment by jackpoz** [comment]
PR updated ! 💯

**Comment by Killyana** [comment]
I tested it again, every thing looks fine.

File: src/server/game/Entities/Creature/GossipDef.cpp [hunk 2]
```diff
@@ -224,10 +224,13 @@ void PlayerMenu::SendGossipMenu(uint32 titleTextId, ObjectGuid objectGUID)
         {
             ++count;
             data << uint32(questID);
-            data << uint32(item.QuestIcon);
+            data << uint32(quest->IsAutoComplete() ? 0 : item.QuestIcon);
```
Comment (Shauren) [inline src/server/game/Entities/Creature/GossipDef.cpp]: I think this is unneccessary - 3.3.5 client only checks 2 values
if QuestIcon == 3 || QuestIcon == 4 then quest is treated as "active"
if QuestIcon != 3 && QuestIcon != 4 then quest is treated as "available"

File: src/server/game/Entities/Creature/GossipDef.cpp [hunk 3]
```diff
@@ -224,10 +224,13 @@ void PlayerMenu::SendGossipMenu(uint32 titleTextId, ObjectGuid objectGUID)
...
             data << int32(quest->GetQuestLevel());
             data << uint32(quest->GetFlags());              // 3.3.3 quest flags
-            data << uint8(0);                               // 3.3.3 changes icon: blue question or yellow exclamation
+            uint8 changeIcon = 0;
+            if (quest->IsAutoComplete() && quest->IsRepeatable())
+                changeIcon = 1;
+            data << uint8(changeIcon);                      // 3.3.3 changes icon: blue question or yellow exclamation
```
Comment (Shauren) [inline src/server/game/Entities/Creature/GossipDef.cpp]: Partial code from 3.3.5 GossipFrame.lua 
```
GossipFrameAvailableQuestsUpdate()
...
		if ( isDaily ) then
			titleButtonIcon:SetTexture("Interface\\GossipFrame\\DailyQuestIcon"); -- blue !
		elseif ( isRepeatable ) then
			titleButtonIcon:SetTexture("Interface\\GossipFrame\\DailyActiveQuestIcon"); -- blue ?
		else
			titleButtonIcon:SetTexture("Interface\\GossipFrame\\AvailableQuestIcon"); -- yellow !
		end
```
isDaily = QuestFlags & 0x1000
isRepeatable = this field

Comment (Shauren) [reply src/server/game/Entities/Creature/GossipDef.cpp]: Active quests (QuestIcon == 3 || QuestIcon == 4) only get 
```
GossipFrameActiveQuestsUpdate()
...
		if ( select(i+3, ...) ) then -- is completed
			titleButtonIcon:SetTexture("Interface\\GossipFrame\\ActiveQuestIcon");
		else
			titleButtonIcon:SetTexture("Interface\\GossipFrame\\IncompleteQuestIcon");
		end		
```

**Comment by Shauren** [comment]
Also, this PR forgets to adjust `PlayerMenu::SendQuestGiverQuestList` which sends the same information using a different opcode

**Comment by jackpoz** [comment]
@Wyrserth feel free to apply the changes Shauren wrote about if you have time :P

File: src/server/game/Entities/Creature/GossipDef.cpp (hunk 2 above)
Comment (Wyrserth) [reply src/server/game/Entities/Creature/GossipDef.cpp]: I have tried with values different than 0, but they cause a bug where you can accept the auto-completable quest and it goes in the quest log.

0 seems to be the only value that works here.

**Comment by Wyrserth** [comment]
Seems to work on all the cases I know about (A Mission Statement, the Candy Bucker quests for the Hallow's End game event) without breaking anything (daily quests have a blue !). Please test other cases if you can :)

**Comment by jackpoz** [comment]
@Killyana could you test this PR ?

**Comment by Killyana** [comment]
I found this issue:
image

**Comment by Killyana** [comment]
Everything looks fine now.

**Comment by Wyrserth** [comment]
One bug remains with monthly quests showing ? instead of !, fixing it now.

**Comment by Wyrserth** [comment]
Fixed monthly quests and a gossip icon bug that affected literally [one quest in the whole game](https://www.wowhead.com/quest=12604). Should be fine now :P

File: src/server/game/Entities/Creature/GossipDef.cpp [hunk 4]
```diff
@@ -226,8 +226,8 @@ void PlayerMenu::SendGossipMenu(uint32 titleTextId, ObjectGuid objectGUID)
             data << uint32(questID);
             data << uint32(item.QuestIcon);
             data << int32(quest->GetQuestLevel());
-            data << uint32(quest->GetFlags());              // 3.3.3 quest flags
-            data << uint8(0);                               // 3.3.3 changes icon: blue question or yellow exclamation
+            data << uint32(quest->GetFlags()); // 3.3.3 quest flags
+            data << uint8(quest->IsRepeatable() && !quest->IsDailyOrWeekly() && !quest->IsMonthly()); // 3.3.3 changes icon: blue question or yellow exclamation
```
Comment (illfated) [inline src/server/game/Entities/Creature/GossipDef.cpp]:
```suggestion
            data << uint8(quest->IsRepeatable() && !quest->IsDailyOrWeekly() && !quest->IsMonthly()); // 3.3.3 icon changes: blue question mark or yellow exclamation mark
```

File: src/server/game/Entities/Creature/GossipDef.cpp [hunk 5]
```diff
@@ -377,8 +377,8 @@ void PlayerMenu::SendQuestGiverQuestList(QEmote const& eEmote, const std::string
             data << uint32(questID);
             data << uint32(questMenuItem.QuestIcon);
             data << int32(quest->GetQuestLevel());
-            data << uint32(quest->GetFlags());             // 3.3.3 quest flags
-            data << uint8(0);                               // 3.3.3 changes icon: blue question or yellow exclamation
+            data << uint32(quest->GetFlags()); // 3.3.3 quest flags
+            data << uint8(quest->IsRepeatable() && !quest->IsDailyOrWeekly() && !quest->IsMonthly()); // 3.3.3 changes icon: blue question or yellow exclamation
```
Comment (illfated) [inline src/server/game/Entities/Creature/GossipDef.cpp]:
```suggestion
            data << uint8(quest->IsRepeatable() && !quest->IsDailyOrWeekly() && !quest->IsMonthly()); // 3.3.3 icon changes: blue question mark or yellow exclamation mark
```

**Review by Wyrserth — APPROVED** [review]

**Pull request merged**